pyccl/halos/profiles.py

```diff
@@ -0,0 +1,824 @@
+        * `plaw_fourier`: FFTLog is able to perform more
+          accurate Hankel transforms by prewhitening its arguments
+          (essentially making them flatter over the range of
+          integration to avoid aliasing). This parameter
+          corresponds to a guess of what the tilt of the profile
+          is (i.e. profile(r) = r^tilt), which FFTLog uses to
+          prewhiten it. This parameter is used when computing the
+          real <-> Fourier transforms. The methods
+          `_get_plaw_fourier` allows finer control over this
+          parameter. The default value allows for a slightly faster
+          (but potentially less accurate) FFTLog transform. Some
+          level of experimentation with this parameter is
+          recommended when implementing a new profile.
+          Default value: -1.5.
+        * `plaw_projected`: same as `plaw_fourier` for the
+          calculation of the 2D projected and cumulative density
+          profiles. Finer control can be achieved with the
+          `_get_plaw_projected`. The default value allows for a
+          slightly faster (but potentially less accurate) FFTLog
+          transform.  Some level of experimentation with this
+          parameter is recommended when implementing a new profile.
+          Default value: -1.
+        """
+        self.precision_fftlog.update(kwargs)
+
+    def _get_plaw_fourier(self, cosmo, a):
+        """ This controls the value of `plaw_fourier` to be used
+        as a function of cosmology and scale factor.
+
+        Args:
+            cosmo (:obj:`Cosmology`): a Cosmology object.
+            a (float): scale factor.
+
+        Returns:
+            float: power law index to be used with FFTLog.
+        """
+        return self.precision_fftlog['plaw_fourier']
+
+    def _get_plaw_projected(self, cosmo, a):
+        """ This controls the value of `plaw_projected` to be
+        used as a function of cosmology and scale factor.
+
+        Args:
+            cosmo (:obj:`Cosmology`): a Cosmology object.
+            a (float): scale factor.
+
+        Returns:
+            float: power law index to be used with FFTLog.
+        """
+        return self.precision_fftlog['plaw_projected']
+
+    def real(self, cosmo, r, M, a, mass_def=None):
+        """ Returns the 3D  real-space value of the profile as a
+        function of cosmology, radius, halo mass and scale factor.
+
+        Args:
+            cosmo (:obj:`Cosmology`): a Cosmology object.
+            r (float or array_like): comoving radius in Mpc.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mass_def (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: halo profile. The shape of the
+            output will be `(N_M, N_r)` where `N_r` and `N_m` are
+            the sizes of `r` and `M` respectively.
```

Comment:
This is still wrong. Sometimes the output shape is simply `(N_r,)` depending on the broadcasting. This shape is different from `(1, N_r,)`. It also not clear if the code would preserve the first dimension. A lot of numpy functions do this extra squeezing. You might find one and follow its conventions for the doc string. Many of them take extra keywords that turn off the squeezing. See `keepdims` on `np.mean` for example: https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html#numpy-mean

Reply:
Right. So the idea here is that, if you pass either M or r/k as a scalar, then that dimension gets squeezed out. I'll document that.

Reply:
OK, how about now?